### ISSUE TITLE
Connection.closed is insufficient

### ISSUE THREAD

**Issue by webloft** [mysql-d/mysql-native#260]
I'm checking `closed` on every query I make on a Windows socket. 

But when the connection was closed by the server the function still returns false and I'm getting a `std.socket.Socket.ERROR` instead.

Edit: Also 
```d
// --- sockets.d:90
enforce!MYX(len != socket.ERROR, "Received std.socket.Socket.ERROR");
```
is not very helpful to set a response to this failure

**Comment by schveiguy** [comment]
> ```d
> sockets.d:90
> ```

So that is where you get the error from?

I agree that the library should properly manage the connection, it already should be doing this.

When you say "Windows socket", what does that mean? You are on Windows?

**Comment by webloft** [comment]
Yes, I'm on Windows using plain phobos socket.

I mean the error isn't helpful because I cannot handle it without a code, I can only compare the error message.

I think the actual bug is in the phobos `Socket.isAlive` itself. I have experienced that it isn't able to proper recognize a closed socket.

**Comment by schveiguy** [comment]
Regardless of how/when it's recognized, there's always a possibility of a race condition. If you check if it's open it could close before you actually start using it.

The library should be robust enough to deal with this.

I will take a look and see if I can figure out a robust way to do this.

**Comment by schveiguy** [comment]
So I just realized that my period vibe-d problem happens because the server has closed the stale connection after some timeout. The next request has a failure to write data.

I'm thinking this is a similar problem for you? What I'm thinking of doing is retrying the connection on the first error without throwing back to the user. Basically, if an error closes the connection, and it was the start of a new query, then retry the connection once.

What do you think? Would this explain the issue you are having?

**Comment by webloft** [comment]
Would fit (for me)

**Comment by schveiguy** [comment]
So the way I'm going to fix this (for now) is to throw a specialized exception `MYXStaleConnection` when a "stale connection" is detected.

This is defined by the first write or the first read failing when the connection was open before starting.

The *hope* is that this situation will not result in an infinite loop of "stale connection" exceptions.

It's kind of sucky that you have to handle this on the outside, but I'm also not sure how much structure to assume to the user's code.

At some point, I want to rewrite the protocol code, and probably encapsulate the sequence of commands a bit better, so I can handle it in the library.

**Comment by schveiguy** [comment]
fixed in #278

**Comment by webloft** [comment]
thx
